src/training/vqa_fine_tune.py

```diff
@@ -0,0 +1,311 @@
+import argparse
+
+import pandas as pd
+import torch
+from PIL import Image
+from torch import nn
+from torch.utils.data import Dataset, DataLoader
+import torch.nn.functional as F
+from tqdm import tqdm
+
+import open_clip
+from open_clip.factory import get_tokenizer
+from training.scheduler import cosine_lr
+from training.train import AverageMeter
+import evaluate
+from sklearn import preprocessing
+import numpy as np
+
+from datasets import load_dataset_builder
+from datasets import load_dataset
+
+class VQATextDataset(Dataset):
+    def __init__(self, df, split, transforms, tokenizer=None):
+        self.df = df
+        self.transforms = transforms
+        self.tokenize = tokenizer
+        self.labels = labelencoder.transform(df['multiple_choice_answer'])
+    def __len__(self):
+        return len(self.df)
+
+    def __getitem__(self, idx):
+        item = self.df.iloc[idx]
+        img_path = item["image"]["path"]
+        image = Image.open(str(img_path))
+        text = item["question"]
+        label = self.labels[idx]
+        return {
+            'image': self.transforms(image),
+            'text': self.tokenize([text])[0],
+            'label': torch.tensor(label)
+        }
+
+def get_task_dataloaders(df, transforms, args):
+    tokenizer = get_tokenizer(args.model)
+
+    dataloaders = {}
+    dataset = VQATextDataset(df,
+        "train",
+        transforms,
+        tokenizer=tokenizer,
+    )
+    dataloader = DataLoader(
+        dataset,
+        batch_size=args.batch_size,
+        shuffle=True,
+        num_workers=args.workers,
+        pin_memory=True,
+        drop_last=True,
+    )
+    dataloaders["train"] = dataloader
+
+    return dataloaders
+
+class CLIPMultimodalClassifier(nn.Module):
+    def __init__(self, encoder, embed_dim, num_labels):
+        super().__init__()
+
+        self.encoder = encoder
+        
+        self.fc1 = nn.Linear(embed_dim * 2, 1536) #size of answer space
```

Comment:
would recommend using `nn.Sequential` like this:

```python
self.mlp = nn.Sequential(
    nn.Linear(embed_dim * 2, 1536),
    nn.GELU(),
    nn.LayerNorm(1536),
    nn.Linear(1536, num_classes),
)
```

and then in the forward:

```
self.mlp(...)
```
rather than doing the whole `self.lnorm(F.relu(...)))`